src/volatile_memory.rs

```diff
@@ -646,38 +740,264 @@ impl<'a, T: ByteValued> VolatileRef<'a, T> {
     ///   let v_ref = unsafe { VolatileRef::<u32>::new(0 as *mut _) };
     ///   assert_eq!(v_ref.len(), size_of::<u32>() as usize);
     /// ```
-    pub fn len(&self) -> usize {
+    pub fn len(self) -> usize {
         size_of::<T>()
     }
 
     /// Does a volatile write of the value `v` to the address of this ref.
     #[inline(always)]
-    pub fn store(&self, v: T) {
+    pub fn store(self, v: T) {
         unsafe { write_volatile(self.addr, Packed::<T>(v)) };
     }
 
     /// Does a volatile read of the value at the address of this ref.
     #[inline(always)]
-    pub fn load(&self) -> T {
+    pub fn load(self) -> T {
         // For the purposes of demonstrating why read_volatile is necessary, try replacing the code
         // in this function with the commented code below and running `cargo test --release`.
         // unsafe { *(self.addr as *const T) }
         unsafe { read_volatile(self.addr).0 }
     }
 
     /// Converts this `T` reference to a raw slice with the same size and address.
-    pub fn to_slice(&self) -> VolatileSlice<'a> {
+    pub fn to_slice(self) -> VolatileSlice<'a> {
         unsafe { VolatileSlice::new(self.addr as *mut u8, size_of::<T>()) }
     }
 }
 
+/// A memory location that supports volatile access of an array of `T`.
+///
+/// # Examples
+///
+/// ```
+/// # use vm_memory::VolatileRef;
+///   let mut v = 5u32;
+///   assert_eq!(v, 5);
+///   let v_ref = unsafe { VolatileRef::<u32>::new(&mut v as *mut u32 as *mut u8) };
+///   assert_eq!(v_ref.load(), 5);
+///   v_ref.store(500);
+///   assert_eq!(v, 500);
+#[derive(Clone, Copy, Debug)]
+pub struct VolatileArrayRef<'a, T: ByteValued>
+where
+    T: 'a,
+{
+    addr: *mut u8,
+    nelem: usize,
+    phantom: PhantomData<&'a T>,
+}
+
+impl<'a, T: ByteValued> VolatileArrayRef<'a, T> {
+    /// Creates a reference to raw memory that must support volatile access of `T` sized chunks.
+    ///
+    /// To use this safely, the caller must guarantee that the memory at `addr` is big enough for
+    /// `nelem` values of type `T` and is available for the duration of the lifetime of the new
+    /// `VolatileRef`. The caller must also guarantee that all other users of the given chunk of
+    /// memory are using volatile accesses.
+    pub unsafe fn new(addr: *mut u8, nelem: usize) -> VolatileArrayRef<'a, T> {
+        VolatileArrayRef {
```

Comment:
Check alignment of addr too? Should we change the return value to Result<VolatileArrayRef<'a, T>>?

Reply:
Same here, `VolatileArrayRef::ref_at` goes through `VolatileRef` which does not have alignment requirements.